cm/server.js

```diff
@@ -0,0 +1,98 @@
+/**
+ * Copyright 2019 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//Insert your values here
+const cmProductToken = 'YOUR_CM_PRODUCT_TOKEN';
+const projectId = 'YOUR_PROJECT_ID';
+
+const express = require('express');
+const request = require('request');
+const messagingApi = require('@cmdotcom/text-sdk');
+const app = express();
+const dialogflowSessionClient = require('../botlib/dialogflow_session_client');
+
+app.use(express.json());
+
+//For authenticating dialogflow_session_client.js, create a Service Account and
+// download its key file. Set the environmental variable
+// GOOGLE_APPLICATION_CREDENTIALS to the key file's location.
+//See https://dialogflow.com/docs/reference/v2-auth-setup and
+// https://cloud.google.com/dialogflow/docs/setup for details.
+
+const cmMessagingApi = new messagingApi.MessageApiClient(cmProductToken);
+const sessionClient = new dialogflowSessionClient(projectId);
+
+const listener = app.listen(process.env.PORT, function () {
+  console.log('Your CM integration server is listening on port '
+    + listener.address().port);
+});
+
+app.post('/', async function (req, res) {
+  const body = req.body;
+  //console.debug("request", JSON.stringify(body));
+  const text = body.message.text ||
+    (!!body.message.media.mediaUri
+      ? (body.message.media.title + '\n media:' + body.message.media.mediaUri)
+      : "unsupported content"
+    );
+  const dialogflowResponse = (await sessionClient.detectIntent(text, body.from.number, body));
+  const response = await sendMessage(dialogflowResponse, body);
+  console.log("response", JSON.stringify(response));
+  return res.send(response);
```

Comment:
May I ask how long CM server will wait for a response?

Reply:
By default 20 seconds. I'd say that should be long enough, no?

Reply:
Actually I'm not super worry about the timeout. I'm more concerned with the following situation. Let's say, a server keeps failing to respond to CM server for some unknown reasons, will CM server block the requests from that server?

Reply:
Messages that fail to get a successful response will be retried every 5 minutes, in theory indefinitely. 

When our 27/4 NOC notices endlessly repeating queues growing too big, they will usually reach out to the specific customer to resolve the issue and/or disable the retry mechanism for the account. (Exact action taken depends on the customer status and account specific agreements.)

So bottom line: any action we will take will be at account or configuration level, never on server or IP level.